### PR TITLE
refactor: (GenAI) Improved Context Cache Samples (Group A)

### PR DESCRIPTION
**Pull request by Thoughtseize1**
## Description
I have modified the following files:

- `create_context_cache.py`
- `delete_context_cache.py`
- `get_context_cache.py`
- `update_context_cache.py`
- `use_context_cache.py`

❓However, as a developer and user, I can not find a single file `list_context_cache.py`

❗I have begun developing this file, but I have encountered an issue: the method required for listing context caches is not yet implemented and cannot be used. I have detailed the problem in the attached [document](https://docs.google.com/document/d/1AvNZYbIhrxO414CmPFRbyL24f-3fpncZQOOaaGy-Kcg/edit?usp=sharing).

Proposed region_tag: `generativeaionvertexai_gemini_list_context_cache`
Could you please review the issue and prioritize the implementation of this method? Having this method available would greatly facilitate the development process.

## Checklist
- [x] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md)
- [ ] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#readme-file)
- [ ] **Tests** pass:   `nox -s py-3.9` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] **Lint** pass:   `nox -s lint` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample
- [ ] Please **merge** this PR for me once it is approved

### CODE REVIEW

File: generative_ai/classify_news_items.py [hunk 1]
```diff
@@ -59,9 +59,16 @@ def classify_news_items() -> str:
     )
 
     print(response.text)
+    """
+    Response example:
```
Comment (msampathkumar) [inline generative_ai/classify_news_items.py:63]:
```suggestion
    Example response:
```

File: generative_ai/classify_news_items.py [hunk 2]
```diff
@@ -59,9 +59,16 @@ def classify_news_items() -> str:
...
+       business
+    """
     # [END generativeaionvertexai_classification]
 
     return response.text
 
 
 # [END aiplatform_sdk_classify_news_items]
```
Comment (msampathkumar) [inline generative_ai/classify_news_items.py:71]: Lets remove region tags for `aiplatform_sdk_classify_news_items`

File: generative_ai/context_caching/update_context_cache.py [hunk 3]
```diff
@@ -36,3 +36,10 @@ def update_context_cache(project_id: str, cache_id: str) -> str:
     # [END generativeaionvertexai_gemini_update_context_cache]
 
     return cached_content.expire_time
+
+
+if __name__ == "__main__":
+    import os
```
Comment (msampathkumar) [inline generative_ai/context_caching/update_context_cache.py:42]: Please move `import` and `PROJECT_ID` to the top portion of the file.

File: generative_ai/context_caching/create_context_cache.py [hunk 4]
```diff
@@ -54,3 +54,10 @@ def create_context_cache(project_id: str) -> str:
     # [END generativeaionvertexai_gemini_create_context_cache]
 
     return cached_content.name
+
+
+if __name__ == "__main__":
+    import google.auth
+
+    project_id = google.auth.default()[1]
```
Comment (msampathkumar) [inline generative_ai/context_caching/create_context_cache.py:62]: Please use `import os` and `PROJECT_ID=os.get...` & move them to the top portion of the file

File: generative_ai/context_caching/get_context_cache.py [hunk 5]
```diff
@@ -31,3 +31,10 @@ def get_context_cache(project_id: str, cache_id: str) -> str:
     # [END generativeaionvertexai_gemini_get_context_cache]
 
     return cached_content.name
+
+
+if __name__ == "__main__":
+    import os
+
+    PROJECT_ID = os.getenv("GOOGLE_CLOUD_PROJECT")
```
Comment (msampathkumar) [inline generative_ai/context_caching/get_context_cache.py:39]: Please use `import os` and `PROJECT_ID=os.get...` & move them to the top portion of the file

File: generative_ai/classify_news_items.py (hunk 2 above)
Comment (Thoughtseize1) [reply generative_ai/classify_news_items.py:71]: Done

File: generative_ai/context_caching/update_context_cache.py (hunk 3 above)
Comment (Thoughtseize1) [reply generative_ai/context_caching/update_context_cache.py:42]: Done

File: generative_ai/context_caching/create_context_cache.py [hunk 6]
```diff
@@ -54,3 +57,7 @@ def create_context_cache(project_id: str) -> str:
     # [END generativeaionvertexai_gemini_create_context_cache]
```
Comment (msampathkumar) [inline generative_ai/context_caching/create_context_cache.py:57]: Can we have `# Example response` here ?

File: generative_ai/context_caching/get_context_cache.py [hunk 7]
```diff
@@ -11,23 +11,30 @@
...
-    # cache_id = "CACHE_ID"
+    # PROJECT_ID = "your-project-id"
+    # cache_id = "your-cache-id"
 
-    vertexai.init(project=project_id, location="us-central1")
+    vertexai.init(project=PROJECT_ID, location="us-central1")
 
     cached_content = caching.CachedContent(cached_content_name=cache_id)
 
     print(cached_content.name)
```
Comment (msampathkumar) [inline generative_ai/context_caching/get_context_cache.py:33]: Can we have example response here ?

File: generative_ai/context_caching/update_context_cache.py [hunk 8]
```diff
@@ -11,28 +11,36 @@
...
-
     cached_content.refresh()
+
     print(cached_content.expire_time)
-    # [END generativeaionvertexai_gemini_update_context_cache]
 
     return cached_content.expire_time
+
+
+# [END generativeaionvertexai_gemini_update_context_cache]
```
Comment (msampathkumar) [inline generative_ai/context_caching/update_context_cache.py:43]: Please move the region tag back to its initial location & also add example reponse

File: generative_ai/context_caching/use_context_cache.py [hunk 9]
```diff
@@ -36,3 +39,7 @@ def use_context_cache(project_id: str, cache_id: str) -> str:
     # [END generativeaionvertexai_gemini_use_context_cache]
```
Comment (msampathkumar) [inline generative_ai/context_caching/use_context_cache.py:39]: Please add example response

File: generative_ai/context_caching/create_context_cache.py (hunk 6 above)
Comment (Thoughtseize1) [reply generative_ai/context_caching/create_context_cache.py:57]: Added

File: generative_ai/context_caching/update_context_cache.py (hunk 8 above)
Comment (Thoughtseize1) [reply generative_ai/context_caching/update_context_cache.py:43]: Done✅

File: generative_ai/context_caching/get_context_cache.py (hunk 7 above)
Comment (Thoughtseize1) [reply generative_ai/context_caching/get_context_cache.py:33]: Added one more print and example output

File: generative_ai/context_caching/use_context_cache.py (hunk 9 above)
Comment (Thoughtseize1) [reply generative_ai/context_caching/use_context_cache.py:39]: Added

File: generative_ai/context_caching/use_context_cache.py [hunk 10]
```diff
@@ -33,6 +36,14 @@ def use_context_cache(project_id: str, cache_id: str) -> str:
...
+    # The provided text is about a new family of multimodal models called Gemini, developed by Google.
+    # ...
+
     # [END generativeaionvertexai_gemini_use_context_cache]
 
     return response.text
+
+
+if __name__ == "__main__":
+    use_context_cache("6374036031984893952")
```
Comment (msampathkumar) [inline generative_ai/context_caching/use_context_cache.py:49]: minor-nit:  can you update this to like below to clearly tell that its an example number ? Please do the same for other samples too.


```suggestion
    use_context_cache("123456789")
```

Comment (Thoughtseize1) [reply generative_ai/context_caching/use_context_cache.py]: Done

File: generative_ai/context_caching/create_context_cache.py [hunk 11]
```diff
@@ -51,6 +54,12 @@ def create_context_cache(project_id: str) -> str:
     )
 
     print(cached_content.name)
+    # Example response:
+    # 6374036031984893952
```
Comment (msampathkumar) [inline generative_ai/context_caching/create_context_cache.py:58]:
```suggestion
    # 1234567890
```

File: generative_ai/context_caching/get_context_cache.py [hunk 12]
```diff
@@ -11,23 +11,34 @@
...
-    vertexai.init(project=project_id, location="us-central1")
+    vertexai.init(project=PROJECT_ID, location="us-central1")
 
     cached_content = caching.CachedContent(cached_content_name=cache_id)
 
+    print(cached_content.resource_name)
     print(cached_content.name)
-    # [END generativeaionvertexai_gemini_get_context_cache]
+    # Example response:
+    # projects/[PROJECT_ID]/locations/us-central1/cachedContents/6374036031984893952
```
Comment (msampathkumar) [inline generative_ai/context_caching/get_context_cache.py:36]:
```suggestion
    # projects/[PROJECT_ID]/locations/us-central1/cachedContents/1234567890
```

File: generative_ai/context_caching/get_context_cache.py [hunk 13]
```diff
@@ -11,23 +11,34 @@
...
+    # 6374036031984893952
```
Comment (msampathkumar) [inline generative_ai/context_caching/get_context_cache.py:37]:
```suggestion
```

File: generative_ai/context_caching/update_context_cache.py [hunk 14]
```diff
@@ -11,32 +11,37 @@
...
-    cached_content.update(ttl=datetime.timedelta(hours=2))
-
+    # Update the expiration time by 1 hour
+    cached_content.update(ttl=datetime.timedelta(hours=1))
     cached_content.refresh()
+
     print(cached_content.expire_time)
-    # [END generativeaionvertexai_gemini_update_context_cache]
+    # Example response:
+    # 2024-09-11 17:16:45.864520+00:00
```
Comment (msampathkumar) [inline generative_ai/context_caching/update_context_cache.py:41]: Let remove this empty line

File: generative_ai/context_caching/get_context_cache.py (hunk 13 above)
Comment (Thoughtseize1) [reply generative_ai/context_caching/get_context_cache.py:37]: Done

File: generative_ai/context_caching/create_context_cache.py (hunk 11 above)
Comment (Thoughtseize1) [reply generative_ai/context_caching/create_context_cache.py]: Done

File: generative_ai/context_caching/update_context_cache.py (hunk 14 above)
Comment (Thoughtseize1) [reply generative_ai/context_caching/update_context_cache.py]: Done

**Review by msampathkumar — APPROVED** [review]
LGTM

**Pull request merged**